method/bass/bass.r

```diff
@@ -0,0 +1,216 @@
+coord_file <- opt$coordinates
+feature_file <- opt$features
+observation_file <- opt$observations
+
+if (!is.na(opt$neighbors)) {
+  neighbors_file <- opt$neighbors
+}
+
+if (!is.na(opt$matrix)) {
+  matrix_file <- opt$matrix
+  matrix <- as(Matrix::readMM(neighbors_file), "CsparseMatrix")
+}
+
+if (!is.na(opt$dim_red)) {
+  dimred_file <- opt$dim_red
+}
+if (!is.na(opt$image)) {
+  image_file <- opt$image
+}
+if (!is.na(opt$config)) {
+  config_file <- opt$config
+  config <- fromJSON(config_file)
+}
+
+get_SpatialExperiment <- function(
+    feature_file,
+    observation_file,
+    coord_file,
+    matrix_file = NA,
+    reducedDim_file = NA,
+    assay_name = "counts",
+    reducedDim_name = "reducedDim") {
+  rowData <- read.delim(feature_file, stringsAsFactors = FALSE, row.names = 1)
+  colData <- read.delim(observation_file, stringsAsFactors = FALSE, row.names = 1)
+
+  coordinates <- read.delim(coord_file, sep = "\t", row.names = 1)
+  coordinates <- as.matrix(coordinates[rownames(colData), ])
+  coordinates[,c(1:2)] <- as.numeric(coordinates[,c(1:2)])
+    
+  spe <- SpatialExperiment::SpatialExperiment(
+    rowData = rowData, colData = colData, spatialCoords = coordinates
+  )
+
+  if (!is.na(matrix_file)) {
+    assay(spe, assay_name, withDimnames = FALSE) <- as(Matrix::t(Matrix::readMM(matrix_file)), "CsparseMatrix")
+    #assay(spe, "logcounts", withDimnames = FALSE) <- log1p(as(Matrix::t(Matrix::readMM(matrix_file)), "CsparseMatrix"))
+  }
+
+  # Filter features and samples
+  if ("selected" %in% colnames(rowData(spe))) {
+    spe <- spe[as.logical(rowData(spe)$selected), ]
+  }
+  if ("selected" %in% colnames(colData(spe))) {
+    spe <- spe[, as.logical(colData(spe)$selected)]
+  }
+
+  if (!is.na(reducedDim_file)) {
+    dimRed <- read.delim(reducedDim_file, stringsAsFactors = FALSE, row.names = 1)
+    reducedDim(spe, reducedDim_name) <- as.matrix(dimRed[colnames(spe), ])
+  }
+  return(spe)
+}
+
+
+technology <- opt$technology
+n_clusters <- opt$n_clusters
+
+# Seed
+seed <- opt$seed
+set.seed(seed)
+
+## Your code goes here
+R <- n_clusters
+C <- config$C
+init_method <- config$init_method
+beta_method <- config$beta_method
+geneSelect <- config$geneSelect
+scaleFeature <- config$scaleFeature
+
+# SpatialExperiment
+spe <- get_SpatialExperiment(
+    feature_file = feature_file,
+    observation_file = observation_file,
+    coord_file = coord_file,
+    matrix_file = matrix_file#,
+    #reducedDim_file = dimred_file
+)
+
+
+# Set up BASS object
+BASS <- createBASSObject(
+    list("experiment" = SummarizedExperiment::assay(spe, "counts")), 
+    list("experiment" = SpatialExperiment::spatialCoords(spe)),
+    C = C, R = R,
+    beta_method = beta_method, init_method = init_method, 
+    burnin = 2000, nsample = 10000)
+
+# Data pre-processing:
+BASS <- BASS.preprocess(BASS, doLogNormalize = TRUE,
+  geneSelect = geneSelect, nSE = 3000, doPCA = TRUE, 
+  scaleFeature = as.logical(scaleFeature), nPC = 20)
```

Comment:
I would maybe pull out `nPC` into the config to make it future proof.

I assume `nSE` is the number of selected genes? If so this could als go into the config.